spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -347,3 +405,152 @@ def _numba_detect_peak_neg(traces, traces_center, peak_mask, exclude_sweep_size,
+        exclude_sweep_size=5,
+        max_window=7,
+        neighbours_mask=None,
+        device=None,
+    ):
+        """
+        Voltage thresholding detection and deduplication with torch.
+
+        Implementation from Charlie Windolf:
+        https://github.com/cwindolf/spike-psvae/blob/ba0a985a075776af892f09adfd453b8d9db168b9/spike_psvae/detect.py#L350
+
+        Parameters
+        ----------
+        traces : np.array
+            Chunk of traces
+        thresholds : np.array
+            Absolute thresholds by channel
+        peak_sign : str, optional
+            "neg", "pos" or "both", by default "neg"
+        exclude_sweep_size : int, optional
+            How many temporal neighbors to compare with during argrelmin, by default 5
+        max_window : int, optional
+            How many temporal neighbors to compare with during dedup, by default 7
+        neighbor_mask : np.array, optional
+            If given, a matrix with shape (num_channels, num_neighbours) with 
+            neighbour indices for each channel. The matrix needs to be rectangular and
+            padded to num_channels, by default None
+        device : str, optional
+            "cpu", "gpu", or None. If None and cuda is available, "gpu" is selected, by default None
+
+        Returns
+        -------
+        sample_inds, chan_inds
+            1D numpy arrays
+        """
+        # TODO handle GPU-memory at chunk executor level
+        # for now we keep the same batching mechanism from spike_psvae
+        # this will be adjusted based on: num jobs, num gpus, num neighbors
+        MAXCOPY = 8
+
+        num_samples, num_channels = traces.shape
+        if device is None:
+            device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        # -- torch argrelmin
+        if peak_sign == "neg":
+            neg_traces = torch.as_tensor(
+                -traces, device=device, dtype=torch.float
+            )
+        elif peak_sign == "pos":
+            neg_traces = torch.as_tensor(
+                traces, device=device, dtype=torch.float
+            )
+        elif peak_sign == "both":
+            neg_traces = torch.as_tensor(
+                -np.abs(traces), device=device, dtype=torch.float
+            )
+        thresholds_torch = torch.as_tensor(thresholds, device=device, dtype=torch.float)
+
+        max_amps, inds = F.max_pool2d_with_indices(
+            neg_traces[None, None],
+            kernel_size=[2 * exclude_sweep_size + 1, 1],
+            stride=1,
+            padding=[exclude_sweep_size, 0],
+        )
+        max_amps = max_amps[0, 0]
+        inds = inds[0, 0]
+        # torch `inds` gives loc of argmax at each position
+        # find those which actually *were* the max
+        unique_inds = inds.unique()
+        window_max_inds = unique_inds[inds.view(-1)[unique_inds] == unique_inds]
+
+        # voltage threshold (@Charlie, added by channel)
+        thresholds_by_channel = torch.as_tensor(
```

Comment:
I'm wondering, why is the `as_tensor` necessary? `thresholds_torch` `window_max_inds` are both tensors already right?

Reply:
You may be right! I'll test (I think I added this before making the abs_thresholds a tensor at the beginning)

Reply:
Got it! was just curious